src/ptr/CowSlice.zig

```diff
@@ -0,0 +1,270 @@
+
+/// "Copy on write" slice. There are many instances when it is desired to re-use
+/// a slice, but doing so would make it unknown if that slice should be freed.
+/// This structure, in release builds, is the same size as `[]const T`, but
+/// stores one bit for if deinitialziation should free the underlying memory.
+///
+///     const str = CowSlice(u8).initOwned(try alloc.dupe(u8, "hello!"), alloc);
+///     const borrow = str.borrow();
+///     assert(borrow.slice().ptr == str.slice().ptr)
+///     borrow.deinit(alloc); // knows it is borrowed, no free
+///     str.deinit(alloc); // calls free
+///
+/// In a debug build, there are aggressive assertions to ensure unintentional
+/// frees do not happen. But in a release build, the developer is expected to
+/// keep slice owners alive beyond the lifetimes of the borrowed instances.
+///
+/// CowSlice does not support slices longer than `2^(@bitSizeOf(usize)-1)`.
+pub fn CowSliceZ(T: type, comptime sentinel: ?T) type {
+    return struct {
+        /// Pointer to the underlying data. Do not access this directly.
+        ///
+        /// NOTE: `ptr` is const if data is borrowed.
+        ptr: [*]T,
+        flags: packed struct(usize) {
+            len: @Type(.{ .int = .{
+                .bits = @bitSizeOf(usize) - 1,
+                .signedness = .unsigned,
+            } }),
+            is_owned: bool,
+        },
+        debug: if (cow_str_assertions) ?*DebugData else void,
+
+        const Self = @This();
+        pub const Slice = if (sentinel) |z| [:z]const T else []const T;
+        pub const SliceMut = if (sentinel) |z| [:z]T else []T;
+
+        pub const empty: Self = initStatic("");
+
+        /// Create a new Cow that owns its allocation.
+        ///
+        /// `data` is transferred into the returned string, and must be freed with
+        /// `.deinit()` when the string and its borrows are done being used.
+        pub fn initOwned(data: []T, allocator: Allocator) Self {
+            if (AllocationScope.downcast(allocator)) |scope|
+                scope.assertOwned(data);
+
+            return .{
+                .ptr = data.ptr,
+                .flags = .{ .is_owned = true, .len = @intCast(data.len) },
+                .debug = if (comptime cow_str_assertions)
+                    bun.new(DebugData, .{
+                        .allocator = allocator,
+                    }),
+            };
+        }
+
+        /// Create a new Cow that copies `data` into a new allocation.
+        pub fn initDupe(data: Slice, allocator: Allocator) !Self {
+            const bytes: Slice = if (comptime sentinel) |_|
+                try allocator.dupeZ(T, data)
+            else
+                try allocator.dupe(T, data);
+
+            return initOwned(bytes, allocator);
+        }
+
+        /// Create a Cow that wraps a static string.
+        ///
+        /// Calling `.deinit()` is safe to call, but will will have no effect.
+        pub fn initStatic(comptime data: Slice) Self {
```

Comment:
this param does not need to be `comptime`. though i see why. the implementation is a copy of `.initUnchecked(data, false)`

Reply:
IMO we should remove `initUnchecked`. Only accepting comptime strings prevents unsafe borrows that may lead to UAFs or leaks.